### PR TITLE
Refactor parser.

### PR DESCRIPTION
**Pull request by chriseth**
This is working towards https://github.com/ethereum/solidity/issues/3520

### CODE REVIEW

File: libsolidity/parsing/Parser.h [hunk 1]
```diff
@@ -145,21 +145,25 @@ class Parser: public ParserBase
...
 	LookAheadInfo peekStatementType() const;
+	/// @returns an IndexAccessedPath as a prestage to parsing a variable declaration (type name)
+	/// or an expression;
+	IndexAccessedPath parseIndexAccessedPath();
 	/// @returns a typename parsed in look-ahead fashion from something like "a.b[8][2**70]".
-	ASTPointer<TypeName> typeNameIndexAccessStructure(
-		std::vector<ASTPointer<PrimaryExpression>> const& _path,
-		std::vector<std::pair<ASTPointer<Expression>, SourceLocation>> const& _indices
-	);
+	ASTPointer<TypeName> typeNameFromIndexAccessStructure(IndexAccessedPath const& _pathAndIndices);
```
Comment (erak) [inline libsolidity/parsing/Parser.h:164]: Perhaps name this function just ``typeName`` or ``retrieveTypeName`` ? The new signature already states the remaining ``fromIndexAccessStructure``. Or is there a semantic difference between ``path`` and ``structure``?

File: libsolidity/parsing/Parser.h [hunk 2]
```diff
@@ -145,21 +145,25 @@ class Parser: public ParserBase
...
 	/// @returns an expression parsed in look-ahead fashion from something like "a.b[8][2**70]".
-	ASTPointer<Expression> expressionFromIndexAccessStructure(
-		std::vector<ASTPointer<PrimaryExpression>> const& _path,
-		std::vector<std::pair<ASTPointer<Expression>, SourceLocation>> const& _indices
-	);
+	ASTPointer<Expression> expressionFromIndexAccessStructure(IndexAccessedPath const& _pathAndIndices);
```
Comment (erak) [inline libsolidity/parsing/Parser.h:166]: Same as above. Perhaps just ``expression`` or ``retrieveExpression``?

File: libsolidity/parsing/Parser.cpp [hunk 3]
```diff
@@ -1098,46 +1098,17 @@ ASTPointer<Statement> Parser::parseSimpleStatement(ASTPointer<ASTString> const&
 	default:
 		break;
 	}
+
 	// At this point, we have 'Identifier "["' or 'Identifier "." Identifier' or 'ElementoryTypeName "["'.
-	// We parse '(Identifier ("." Identifier)* |ElementaryTypeName) ( "[" Expression "]" )+'
+	// We parse '(Identifier ("." Identifier)* |ElementaryTypeName) ( "[" Expression "]" )*'
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp:1103]: Is this typo there in the grammar too?

**Review by axic — APPROVED** [review]

File: libsolidity/parsing/Parser.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:1103]: I think the grammar does not have that form.

**Pull request merged**